nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GenerateTableFetch.java

```diff
@@ -0,0 +1,254 @@
+
+        // Build a SELECT query with maximum-value columns (if present)
+        final String selectQuery = dbAdapter.getSelectStatement(tableName, columnsClause, whereClause, null, null, null);
+
+        try (final Connection con = dbcpService.getConnection();
+             final Statement st = con.createStatement()) {
+
+            final Integer queryTimeout = context.getProperty(QUERY_TIMEOUT).asTimePeriod(TimeUnit.SECONDS).intValue();
+            st.setQueryTimeout(queryTimeout); // timeout in seconds
+
+            logger.debug("Executing {}", new Object[]{selectQuery});
+            ResultSet resultSet = st.executeQuery(selectQuery);
+            if (resultSet.next()) {
+                // Total row count is in the first column
+                int rowCount = resultSet.getInt(1);
+
+                // Update the state map with the newly-observed maximum values
+                ResultSetMetaData rsmd = resultSet.getMetaData();
+                for (int i = 2; i <= rsmd.getColumnCount(); i++) {
+                    String resultColumnName = rsmd.getColumnName(i).toLowerCase();
+                    int type = rsmd.getColumnType(i);
+                    try {
+                        String newMaxValue = getMaxValueFromRow(resultSet, i, type, statePropertyMap.get(resultColumnName.toLowerCase()), dbAdapter.getName());
+                        if (newMaxValue != null) {
+                            statePropertyMap.put(resultColumnName, newMaxValue);
+                        }
+                    } catch (ParseException | IOException pie) {
+                        // Fail the whole thing here before we start creating flow files and such
+                        throw new ProcessException(pie);
+                    }
+                }
+
+                final int numberOfFetches = (partitionSize == 0) ? rowCount : (rowCount / partitionSize) + (rowCount % partitionSize == 0 ? 0 : 1);
+
+                // Generate SQL statements to read "pages" of data
+                for (int i = 0; i < numberOfFetches; i++) {
+                    FlowFile sqlFlowFile = null;
+                    try {
+                        Integer limit = partitionSize == 0 ? null : partitionSize;
+                        Integer offset = partitionSize == 0 ? null : i * partitionSize;
+                        final String query = dbAdapter.getSelectStatement(tableName, columnNames, StringUtils.join(maxValueClauses, " AND "), null, limit, offset);
+                        sqlFlowFile = session.create();
+                        sqlFlowFile = session.write(sqlFlowFile, out -> {
+                            out.write(query.getBytes());
+                        });
+                        session.transfer(sqlFlowFile, REL_SUCCESS);
+
+                    } catch (Exception e) {
+                        logger.error("Error while generating SQL statement", e);
+                        if (sqlFlowFile != null) {
+                            session.remove(sqlFlowFile);
```

Comment:
The session may fail to write the contents of one flowfile but succeed for another for one reason or another. If any fail, I'd suggest wrapping the for loop with the try/catch so it will stop processing and roll back the whole session so no data is lost.